### PR TITLE
Integrate `gohashtree` hashing library

### PR DESCRIPTION
**Pull request by rkapka**
This PR allows using https://github.com/prysmaticlabs/gohashtree for hashing during merkleization. The algorithm used can be toggled through the `EnableVectorizedHTR` public variable.

### CODE REVIEW

File: hasher.go [hunk 1]
```diff
@@ -415,3 +483,88 @@ func (h *Hasher) merkleizeImpl(dst []byte, input []byte, limit uint64) []byte {
 	}
 	return append(dst, res...)
 }
+
+func merkleizeInput(input []byte, limit uint64) []byte {
+	elemCount := len(input) / 32
```
Comment (potuz) [inline hasher.go:488]: Don't you have an off-by one here? if `len(input) = 1` don't you want to have one chunk? I would have expected here something like `(len(input) + 31)/32`

File: hasher.go [hunk 2]
```diff
@@ -415,3 +483,88 @@ func (h *Hasher) merkleizeImpl(dst []byte, input []byte, limit uint64) []byte {
...
+	elements := make([][32]byte, elemCount)
+	elemLen := len(elements)
```
Comment (potuz) [inline hasher.go:490]: why this variable instead of using `elemCount`?

File: hasher.go [hunk 3]
```diff
@@ -415,3 +483,88 @@ func (h *Hasher) merkleizeImpl(dst []byte, input []byte, limit uint64) []byte {
...
+	}
+	for i := uint8(0); i < dep; i++ {
+		layerLen := len(elements)
+		oddNodeLength := layerLen%2 == 1
+		if oddNodeLength {
+			zerohash := zeroHashesRaw[i]
+			elements = append(elements, zerohash)
+		}
+		outputLen := len(elements) / 2
+		err := gohashtree.Hash(elements, elements)
```
Comment (potuz) [inline hasher.go:525]: Perhaps adding comment here that gohashtree overwrites concurrently elements is worth it.

File: hasher.go [hunk 4]
```diff
@@ -415,3 +483,88 @@ func (h *Hasher) merkleizeImpl(dst []byte, input []byte, limit uint64) []byte {
...
+		if err != nil {
+			panic(err)
+		}
+		elements = elements[:outputLen]
+	}
+	return elements[0]
+}
+
+// Depth retrieves the appropriate depth for the provided trie size.
+func depth(v uint64) (out uint8) {
```
Comment (potuz) [inline hasher.go:535]: Can you add the test for this one? I think it was in proto's repo right?

Comment (rkapka) [reply hasher.go:535]: I found this in our repo and it has only this test:
```
func TestDepthOfOne(t *testing.T) {
	assert.Equal(t, uint8(0), ssz.Depth(1))
}
```
Proto's repo has no tests: https://github.com/protolambda/zssz/tree/master/merkle

**Review by potuz — APPROVED** [review]

**Pull request merged**